pennylane_lightning_kokkos/src/simulator/StateVectorKokkos.hpp

```diff
@@ -445,6 +447,149 @@ template <class Precision> class StateVectorKokkos {
+                            cld[idx] = probabilities[idx];
+                        else
+                            cld[idx] = probabilities[idx] + cld[idx - 1];
+                    }
+                }
+            });
+
+        for (size_t i = 1; i < Nsqrt; i++) {
+            Kokkos::parallel_for(
+                Kokkos::RangePolicy<KokkosExecSpace>(0, Nsqrt),
+                KOKKOS_LAMBDA(const size_t &j) {
+                    size_t idx = i * Nsqrt + j;
+                    size_t idx0 = i * Nsqrt - 1;
+                    if (idx < N)
+                        cld[idx] += cld[idx0];
+                });
+        }
+
+        Kokkos::Random_XorShift64_Pool<> rand_pool(5374857);
+
+        Kokkos::parallel_for(
+            Kokkos::RangePolicy<KokkosExecSpace>(0, num_samples),
+            KOKKOS_LAMBDA(const size_t &i) {
+                Kokkos::Random_XorShift64_Pool<>::generator_type rand_gen =
+                    rand_pool.get_state();
+
+                double U = rand_gen.drand(0.0, 1.0);
+
+                size_t idx;
+
+                if (U <= cld[0]) {
+                    idx = 0;
+                } else {
+                    size_t lo = 0, hi = N - 1;
+                    size_t mid;
+                    while (hi - lo > 1) {
+                        mid = (hi + lo) / 2;
+                        if (cld[mid] < U)
+                            lo = mid;
+                        else
+                            hi = mid;
+                    }
+                    idx = hi;
+                }
+                for (size_t j = 0; j < num_qubits; j++) {
+                    samples[i * num_qubits + (num_qubits - 1 - j)] =
+                        (idx >> j) & 1U;
+                }
+                rand_pool.free_state(rand_gen);
+            });
+        */
+
+        Kokkos::View<Kokkos::complex<Precision> *> arr_data = getData();
+
+        Kokkos::View<Precision *> probabilities("probabilities", N);
+
+        Kokkos::View<Precision *> cdf("cdf", N);
+
+        Kokkos::View<size_t *> samples("num_samples", num_samples * num_qubits);
+
+        // Compute prob
+        Kokkos::parallel_for(
+            Kokkos::RangePolicy<KokkosExecSpace>(0, N),
+            getProbFunctor<Precision>(arr_data, probabilities));
+
+        // Compute local CDF: each thread compute a chunk of local CDF
+        Kokkos::parallel_for(
+            Kokkos::RangePolicy<KokkosExecSpace>(0, Nsqrt),
+            getLocalCDFFunctor<Precision>(probabilities, cdf, Nsqrt, N));
+
+        Kokkos::fence();
+        // Convert local CDF to global CDF: each thread compute one element
+        // of a chunk
+        for (size_t i = 1; i < Nsqrt; i++) {
+            Kokkos::parallel_for(
+                Kokkos::RangePolicy<KokkosExecSpace>(0, Nsqrt),
+                getGlobalCDFFunctor<Precision>(cdf, i, Nsqrt, N));
+        }
+
+        Kokkos::fence();
+        // Sampling process
+        Kokkos::Random_XorShift64_Pool<> rand_pool(5374857);
```

Comment:
Any reason why the `Random_XorShift64_Pool` instead of the `Random_XorShift1024_Pool`?

Reply:
Good question! Generally speaking, the performance of `Random_XorShift64_Pool` is comparable with that of `Random_XorShift1024_Pool`.  `Random_XorShift64_Pool` always ranks 5th, while `Random_XorShift1024_Pool` always ranks 6th as shown [here](https://arxiv.org/pdf/1811.04035.pdf). `Random_XorShift64_Pool` is recommended if we need consider the memory access cost as illustrated [here](https://arxiv.org/pdf/1402.6246.pdf). There is evidence that `Random_XorShift64_Pool` converges faster as shown in [Fig.9 ](https://arxiv.org/pdf/1402.6246.pdf).